### ISSUE TITLE
[Feature request] access handler response in middleware

### ISSUE THREAD

**Issue by tao1** [actions-on-google/actions-on-google-nodejs#182]
Hello,

I want to integrate chatbase to my dialogflow bot. Notice I use actions-on-google version 2.1.3. Today I can use middleware to intercept requests but not responses forged by my intent handlers. I would be very convenient to send bot responses to chatbase in generic way.

Kind regards,

Laurent.

**Comment by marionfi** [comment]
+1

**Comment by meelie** [comment]
+1

**Comment by igilham** [comment]
I agree. It would be useful to insert middleware before the response is returned.

**Comment by afirstenberg** [comment]
There are two locations where it would be interesting to hook in. Before a response is set (to either add something standard common for multiple Intents, or to have the handlers set values and have the response come from a single function) and after it is sent (to log the interaction to Chatbase, Dashbot, or similar).

**Comment by Canain** [comment]
We have talked about creating a `preware` and `postware` system similar to middlewares right now.

Will need to figure out the exact details, but ideally it can cover most use cases and intercept requests to modify them coming in and intercept responses coming out.

**Comment by igilham** [comment]
This came up again for my team recently. We want a hook to perform analytics reporting after handling the intent but before returning the response. The `postware` idea would make this easy.

It might be worth looking at the Alexa SDK for inspiration. They already support pre- and post- intent handling interceptors, and a more scalable intent handler mechanism.

**Comment by MatheusAnciloto** [comment]
Any updates on this issue?

**Comment by Fleker** [comment]
It has not been something that I've been looking at. Any such feature would also be mirrored in the `@assistant/conversation` library as well.

You can accomplish something similar to what is done in the [dialogflow-session-entities-plugin](https://github.com/actions-on-google/dialogflow-session-entities-plugin-nodejs/blob/master/src/index.ts#L60) manually.

```typescript
  send() {
    const responseBody = this.conv.serialize() as ResponseBody
    const convBody = this.conv.body as GoogleCloudDialogflowV2WebhookRequest
    responseBody.sessionEntityTypes = this.entities.map(entity => {
      entity.name = `${convBody.session}/entityTypes/${entity.name}`
      return entity
    })
    this.conv.json(responseBody)
  }
```

I realize a `postware` method would be easier, but want to make sure I collect enough use-cases so a final library API is suitably flexible. (And whether a `preware` makes sense)
